execute/internal/cache/commit_report_cache.go

```diff
@@ -0,0 +1,190 @@
+	"sort"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+
+	readerpkg "github.com/smartcontractkit/chainlink-ccip/pkg/reader"
+	"github.com/smartcontractkit/chainlink-ccip/pkg/types/ccipocr3"
+)
+
+// CommitReportsCache optimizes calls to CommitReportsGTETimestamp by caching reports
+// and only fetching new reports since the last successful query.
+type CommitReportsCache struct {
+	lggr               logger.Logger
+	reports            *cache.Cache // key string -> CommitPluginReportWithMeta
+	lastQueryTimestamp time.Time    // timestamp of the most recent query
+}
+
+// NewCommitReportsCache creates a new commit reports cache.
+// Cached items live for messageVisibilityInterval + EvictionGracePeriod, mirroring
+// the lifetime used in commitRootsCache for executed roots.
+func NewCommitReportsCache(lggr logger.Logger, messageVisibilityInterval time.Duration) *CommitReportsCache {
+	ttl := messageVisibilityInterval + EvictionGracePeriod
+	return &CommitReportsCache{
+		lggr:    lggr,
+		reports: cache.New(ttl, CleanupInterval),
+	}
+}
+
+// generateKey creates a unique key for a commit report
+func generateKey(report ccipocr3.CommitPluginReportWithMeta) string {
+	// We'll create a unique key for each report by combining relevant fields
+	// For each blessed merkle root
+	for _, mrc := range report.Report.BlessedMerkleRoots {
+		return fmt.Sprintf("%s_%s", mrc.ChainSel.String(), mrc.MerkleRoot.String())
+	}
+
+	// If no blessed roots (unlikely but possible), check unblessed roots
+	for _, mrc := range report.Report.UnblessedMerkleRoots {
+		return fmt.Sprintf("%s_%s", mrc.ChainSel.String(), mrc.MerkleRoot.String())
+	}
+
+	// Fallback if no merkle roots (should never happen)
+	return fmt.Sprintf("%d_%d", report.Timestamp.Unix(), report.BlockNum)
+}
+
+// GetCachedAndNewReports combines cached reports with newly fetched ones
+func (c *CommitReportsCache) GetCachedAndNewReports(
+	ctx context.Context,
+	ccipReader readerpkg.CCIPReader,
+	fetchFrom time.Time,
+	limit int,
+) ([]ccipocr3.CommitPluginReportWithMeta, error) {
+	// Start with cached reports that are newer than or equal to fetchFrom
+	cachedReports := c.getCachedReports(fetchFrom)
+
+	// Determine if we need to fetch new reports
+	var queryFrom time.Time
+	// No lock required for reports (go-cache is thread-safe), but we still protect lastQueryTimestamp.
+	if c.lastQueryTimestamp.IsZero() {
+		// First query, use the original fetchFrom
+		queryFrom = fetchFrom
+	} else {
+		// For subsequent queries, start from the last query timestamp
+		// This is the key optimization: we only query for new reports since our last query
+		queryFrom = c.lastQueryTimestamp
+	}
+
+	// Fetch finalized reports since last query and update the cache with them.
+	newFinalizedReports, err := ccipReader.CommitReportsGTETimestamp(ctx, queryFrom, primitives.Finalized, limit)
```

Comment:
Check if the cache already has enough pending reports to reach the limit?

Might also consider 2 limits:
1. The query limit when fetching the mix of roots & prices.
2. Number of pending commit reports.